### PR TITLE
#20 Add follower/following functionality

### PR DESCRIPTION
**Pull request by Mywe4ka**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -7,6 +8,10 @@ def current_user
     @current_user ||= User.find(session[:user_id]) if session[:user_id]
   end
 
+  def current_user= (user)
```
Comment (VeronikaSalo) [inline app/controllers/application_controller.rb]: Please remove this hack )

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def signed_in?
+    !current_user.nil?
+  end
+
+  def signed_in_user
+    unless signed_in?
+      store_location
+      redirect_to root_url, notice: "Please sign in."
```
Comment (VeronikaSalo) [inline app/helpers/sessions_helper.rb]: Add to I18n

File: app/models/user.rb [hunk 3]
```diff
@@ -27,6 +33,18 @@ def self.from_omniauth(auth)
     end
   end
 
+  def following?(other_user)
+    relationships.find_by_followed_id(other_user.id)
```
Comment (VeronikaSalo) [inline app/models/user.rb]: Please change `find_by...` to `where`

File: app/models/user.rb [hunk 4]
```diff
@@ -27,6 +33,18 @@ def self.from_omniauth(auth)
...
+  end
+
+  def follow!(other_user)
+    relationships.create!(followed_id: other_user.id)
+  end
+
+  def unfollow!(other_user)
+    relationships.find_by_followed_id(other_user.id).destroy
```
Comment (VeronikaSalo) [inline app/models/user.rb]: Here as well. (See comment above)

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -9,7 +9,7 @@
 <body>
   <div id="user_nav">
     <% if current_user %>
-      Signed in as <strong><%= current_user.name %></strong>!
+      Signed in as <strong><%= link_to current_user.name, current_user %></strong>!
```
Comment (VeronikaSalo) [inline app/views/layouts/application.html.erb:12]: Move to I18n as well

File: app/views/users/_stats.html.haml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+- @user ||= current_user
+.stats
+  %a{:href => following_user_path(@user)}
+    %strong#following.stat
+      = @user.followed_users.count
+    following
+  %a{:href => followers_user_path(@user)}
+    %strong#followers.stat
+      = @user.followers.count
+    followers
```
Comment (VeronikaSalo) [inline app/views/users/_stats.html.haml:10]: I18n

File: app/views/users/show.html.haml [hunk 7]
```diff
@@ -6,4 +6,9 @@
 = @user.name
 %br/
 %br/
-= link_to 'Edit', edit_user_path(@user)
\ No newline at end of file
+= render 'follow_form' if signed_in?
```
Comment (VeronikaSalo) [inline app/views/users/show.html.haml]: Put code under one signed_in?

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -25,4 +33,12 @@
     expect( user.name ).to eq(user.first_name.to_s + ' ' + user.last_name.to_s)
   end
 
+  describe "following" do
```
Comment (VeronikaSalo) [inline spec/models/user_spec.rb]: Where the tests itself?

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -19,6 +20,20 @@ def update
     end
   end
 
+  def following
+    @title = "Following"
```
Comment (VeronikaSalo) [inline app/controllers/users_controller.rb]: move it to view and create one _form partioa which will be used in both: followin and followers views

**Comment by VeronikaSalo** [comment]
@Mywe4ka is it ready to merge?

**Comment by Mywe4ka** [comment]
@VeronikaSalo  yes

**Pull request merged**